### PR TITLE
Zip 216 fix

### PR DESCRIPTION
**Pull request by CPerezz**
Since we're not in prod yet, we don't need to be backwards compatible
with the previous encoding and simply accept the next one.

The solution prevents a point of order 2 (0, -1) to be accepted as a
valid point.

For more details, see: https://zips.z.cash/zip-0216

Special thanks to Zcash and @str4d for pinging and filing an issue in
regards of that.

Resolves:  #82

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -523,7 +527,16 @@ impl Serializable<32> for JubJubAffine {
...
+                // the encoding as non-canonical if all of the
+                // following occur:
+                // - x == 0
+                // - flip_sign == true
+                let x_is_zero = x.ct_eq(&BlsScalar::zero());
+                CtOption::new(
+                    JubJubAffine { x: final_x, y },
+                    !(x_is_zero & flip_sign),
+                )
             }),
```
Comment (ZER0) [inline src/lib.rs:540]: Nit: I would avoid to create `x_negated` and `final_x`, the first can just be `&-x` and the second can be just `x`. That would also simplify the creation of the `JubJubAffine`.
In short, something like:
```rust
                // If x == 0, flip_sign == sign_bit. We therefore want to reject
                // the encoding as non-canonical if all of the
                // following occur:
                // - x == 0
                // - flip_sign == true
                let x_is_zero = x.ct_eq(&BlsScalar::zero());
                // Fix the sign of `x` if necessary
                let flip_sign = Choice::from((x.to_bytes()[0] ^ sign) & 1);
                let x = BlsScalar::conditional_select(&x, &-x, flip_sign);

                CtOption::new(JubJubAffine { x, y }, !(x_is_zero & flip_sign))
```

File: src/lib.rs [hunk 2]
```diff
@@ -1581,3 +1594,35 @@ fn test_serialization_consistency() {
...
+        // (0, -1) with sign bit set to 1.
+        [
+            0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff, 0xfe, 0x5b, 0xfe,
+            0xff, 0x02, 0xa4, 0xbd, 0x53, 0x05, 0xd8, 0xa1, 0x09, 0x08, 0xd8,
+            0x39, 0x33, 0x48, 0x7d, 0x9d, 0x29, 0x53, 0xa7, 0xed, 0xf3,
+        ],
+    ];
+
+    for b in &NON_CANONICAL_ENCODINGS {
+        {
```
Comment (ZER0) [inline src/lib.rs:1616]: why the additional block?

**Review by ZER0 — APPROVED** [review]
LGTM, with a couple of nits addressed.

File: src/lib.rs (hunk 2 above)
Comment (CPerezz) [reply src/lib.rs:1616]: My bad.. Fixed

File: src/lib.rs (hunk 1 above)
Comment (CPerezz) [reply src/lib.rs:540]: Ye, I like it more!! Done!

**Review by Bounce23 — APPROVED** [review]
LGTM

**Pull request merged**